### PR TITLE
feat: filters proposals by label

### PR DESCRIPTION
**Pull request by wa0x6e**
### Summary

Closes: https://github.com/snapshot-labs/workflow/issues/254
Depends on https://github.com/snapshot-labs/sx-monorepo/pull/914

This PR adds a new filter to the proposals page, to filter proposals by one or more labels.

Labels are filtered using an `OR` operator, meaning that if filtering by A and B, proposals with either A or B will be returned (and not proposals having both A and B)

Screenshot 2024-10-21 at 17 16 35

Screenshot 2024-10-21 at 17 16 42

Screenshot 2024-10-21 at 17 16 49

### How to test

1. Go to a space proposals page
2. Try to select some labels in the labels filter
3. It should refresh the proposal list, and show only proposals with the related labels

### Note

- Ability to link individual label to the proposals page will come in a later PR: https://github.com/snapshot-labs/sx-monorepo/pull/924

### CODE REVIEW

**Comment by bonustrack** [comment]
We should also make labels clickable on proposal preview and proposal page and redirect to proposals page with the label filter.

**Comment by bonustrack** [comment]
Can we add the labels on /home page and user page as well? And can we make labels clickable on the single proposal page, (redirecting to /proposals with label selected).

**Comment by wa0x6e** [comment]
> Can we add the labels on /home page and user page as well? And can we make labels clickable on the single proposal page, (redirecting to /proposals with label selected).

Labels already appearing on all proposals list (but notifications)

**Comment by wa0x6e** [comment]
@bonustrack is the current UI ok ? or do you have something better ?

File: apps/ui/src/views/Space/Proposals.vue
Comment (ChaituVR) [inline apps/ui/src/views/Space/Proposals.vue:121]: Not sure if we want to use different layouts with/without labels. @bonustrack could you check it in mobile?

File: apps/ui/src/views/Space/Proposals.vue
Comment (ChaituVR) [inline apps/ui/src/views/Space/Proposals.vue:161]: max width is different based on first label we select, is it expected?
Untitled 5
Untitled 4

File: apps/ui/src/views/Space/Proposals.vue
Comment (ChaituVR) [inline apps/ui/src/views/Space/Proposals.vue:211]: Not important! Just a small UX thought, when we click on clear button the picker is still open
Oct-23-2024 08-20-47

User may click it when they don't want any label filters, showing the modal mean they have to click outside one more time to close the picker

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: apps/ui/src/components/PickerLabel.vue
Comment (wa0x6e) [reply apps/ui/src/components/PickerLabel.vue:57]: I am using same z-index as the state dropdown menu

Screenshot 2024-10-23 at 12 49 22

Will try, maybe z-20 is enough

Comment (wa0x6e) [reply apps/ui/src/components/PickerLabel.vue:57]: The dropdown `9999` is making it appears on top of the topnav on scroll, separate issue 

Screenshot 2024-10-23 at 12 51 17

File: apps/ui/src/views/Space/Proposals.vue
Comment (wa0x6e) [reply apps/ui/src/views/Space/Proposals.vue:121]: In my mind, every mobile layout will use the 2 lines version, once we're adding more filter options, like https://github.com/snapshot-labs/sx-monorepo/issues/460

File: apps/ui/src/views/Space/Proposals.vue
Comment (wa0x6e) [reply apps/ui/src/views/Space/Proposals.vue]: Yes. It will only show the first 2 labels, with a width capped at 230px. If your first 2 labels are very short, it will have different length.

We could also force the filter to have a fixed width of 230px

File: apps/ui/src/views/Space/Proposals.vue
Comment (ChaituVR) [inline apps/ui/src/views/Space/Proposals.vue:206]: maybe
```suggestion
                    title="Clear all labels"
```

File: apps/ui/src/views/Space/Proposals.vue
Comment (ChaituVR) [inline apps/ui/src/views/Space/Proposals.vue:91]: i thought i added a comment in my previous review, but missed it,  can we make it just `labels` or `label` instead?
It looks like this now:
`http://127.0.0.1:8080/#/s:thanku.eth/proposals?labels[]=3fe1a9b9&labels[]=b4258576&labels[]=9ad0740d`
I think should be something like `labels=abc,xyz,etc`

**Review by ChaituVR — CHANGES_REQUESTED** [review]

**Comment by bonustrack** [comment]
We don't need the counter here, event if we improve the style, it's not too relevant, most of time users wont filter 30 labels so we dont really need to count them:
image
Otherwise it looks good

**Comment by wa0x6e** [comment]
> We don't need the counter here, event if we improve the style, it's not too relevant, most of time users wont filter 30 labels so we dont really need to count them, we better save space: image Otherwise it looks good

How do you tell how many labels I have selected in this case ? 

Screenshot 2024-10-26 at 15 12 01

2 or more will have the same UI

**Comment by wa0x6e** [comment]
Last minute performance addition: added a throttle to the watcher, so we "batch" requests when user is selecting multiple labels

**Comment by bonustrack** [comment]
> Last minute performance addition: added a throttle to the watcher, so we "batch" requests when user is selecting multiple labels

Isn't that mostly useful when there can be lot of update going on?

**Comment by wa0x6e** [comment]
> > Last minute performance addition: added a throttle to the watcher, so we "batch" requests when user is selecting multiple labels
> 
> Isn't that mostly useful when there can be lot of update going on?

User can select multiple labels

**Comment by bonustrack** [comment]
But it's not like they typing something, they will already take like half second to select another label

**Comment by wa0x6e** [comment]
Maybe i’m just fast, we could increase throttle, or leave it like that

**Comment by bonustrack** [comment]
Ok, I think we can leave it like this, can't be bad.

**Review by bonustrack — APPROVED** [review]
tACK, just waiting for @Sekhmet to do final review of the code

File: apps/ui/src/views/Space/Proposals.vue
Comment (Sekhmet) [inline apps/ui/src/views/Space/Proposals.vue:40]: Let's use types there:
```diff
diff --git a/apps/ui/src/components/PickerLabel.vue b/apps/ui/src/components/PickerLabel.vue
index f3b6ecf6..be695453 100644
--- a/apps/ui/src/components/PickerLabel.vue
+++ b/apps/ui/src/components/PickerLabel.vue
@@ -16,6 +16,10 @@ const props = defineProps<{
   panelProps?: Record<string, any>;
 }>();
 
+defineSlots<{
+  button(props: { close: () => void }): any;
+}>();
+
 const selectedLabels = defineModel<string[]>({
   required: true
 });
diff --git a/apps/ui/src/views/Space/Proposals.vue b/apps/ui/src/views/Space/Proposals.vue
index ebc86d6a..a8f6448c 100644
--- a/apps/ui/src/views/Space/Proposals.vue
+++ b/apps/ui/src/views/Space/Proposals.vue
@@ -34,7 +34,7 @@ const spaceLabels = computed(() => {
   return Object.fromEntries(props.space.labels.map(label => [label.id, label]));
 });
 
-function handleClearLabelsFilter(close: any) {
+function handleClearLabelsFilter(close: () => void) {
   labels.value = [];
   close();
 }
```

**Review by Sekhmet — COMMENTED** [review]
Other than that it looks good to me.

**Review by ChaituVR — APPROVED** [review]
Just to remove this 
Untitled 4

**Comment by bonustrack** [comment]
@wa0x6e Can we merge this?

**Comment by wa0x6e** [comment]
> @wa0x6e Can we merge this?

Up to the reviewers, for the latest changes

**Comment by bonustrack** [comment]
I believe it's good to go, you already addresses comments from Wik.

**Pull request merged**
